expressions/literals.md

```diff
@@ -0,0 +1,319 @@
+
+If the file containing this code is encoded as *UTF-8* 
+the byte-value of ``u_umlaut`` will be: ``\xc3\xbc``.
+If the file is encoded with *ISO-8559-1* (Latin-1) its value is ``\xfc``.
+
+``String`` works best with *UTF-8* as it has methods to deal with unicode
+codepoints which expects *UTF-8*.
+
+So it is advised to encode you pony source code in *UTF-8*.
+
+### Triple quoted Strings
+
+For embedding multi-line text in string literals,
+there are triple quoted strings.
+
+```pony
+let triple_quoted_string_docs =
+  """
+  Triple quoted strings are the way to go for long multiline text.
+  They are extensively used as docstrings which are turned into api documentation.
+
+  They get some special treatment, in order to keep pony code readable:
+
+  * The string literal starts on the line after the opening triple quote.
+  * Common indentation is removed from the string literal
+    so it can be conveniently aligned with the enclosing indentation
+    e.g. each line of this literal will get its first two whitespaces removed
+  * Whitespace after the opening and before the closing triple quote will be
+    removed as well. The first line will be completely removed.
+    if it only contains whitespace.
+    e.g. this strings first character is ``T`` not ``\n``.
+  """
+```
+
+## Array Literals
+
+Array literals are enclosed by square brackets.
+Array literal elements can be any kind of expressions.
+They are separated by semicolon or newline:
+
+```pony
+let my_literal_array = [ "first"; "second"
+  "third one on a new line"]
+```
+
+Empty array literals are not supported.
+Use ``recover val Array(0) end`` instead.
+
+### Type inference
+
+
+The type of the literal array expression is Always ``Array[T] ref``
+where ``T`` (the type of the elements) is inferred as 
+the union type of all the element types:
+
+```pony
+let my_heterogenous_array: Array[(U64|String)] ref = [
+  U64(42)
+  "42"
+  U64.min_value()
+]
+```
+
+In the above example the resulting array type will be
+``Array[(U64|String)] ref`` because the array contains
+``String`` and ``U64`` elements.
+
+It is possible to give the literal a hint on what kind of type it should
+coerce the array literal to. In the above example
+we might only want an ``Array[Stringable] ref``. ``Stringable`` is a trait
+that both ``String`` and ``U64`` implement.
+
+In order to do so, add an ``as`` Expression
+defining the desired Array element type right after the opening square bracket
+delimited by a colon:
+
+```pony
+let my_stringable_array: Array[Stringable] ref =
+  [as Stringable:
+    U64(0xFFEF)
+    "0xFFEF"
+    U64(1 + 1)
+  ]
+```
+
+This Array literal is coerced to be an ``Array[Stringable] ref`` according to
+the ``as`` expression.
+
+Inferring the element type from the type of the local variable
+or field it is assigned to is not yet possible:
+
+```pony
+// does not compile
+let my_stringable_array: Array[Stringable] =
+  [
+    U64(0xA)
+    "0xA"
+  ]
+```
+
+### Arrays and References
```

Comment:
I do think that this is going into too much detail about this aspect of arrays.

This is not really a property of array literals so much as it being a property of capabilities. If your capability has uniqueness constraints (`iso` or `trn`), then you can't hold it as such in two places at once, including the particular case where one of those places is in an array, but note that the array isn't really a special case for this at all.

I think we can describe everything below this header succinctly in a sentence or two without going through full worked examples about it.

Reply:
I shortened this section to only mention that element types are inferred as their aliases.